src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkWasmModule.java

```diff
@@ -0,0 +1,278 @@
+        try {
+          wasmModule = com.dylibso.chicory.wasm.Parser.parse(moduleContent);
+        } catch (ChicoryException e) {
+          throw new EvalException(e);
+        }
+      }
+      validateModule(wasmModule, allocateFn);
+    }
+
+    this.path = path;
+    this.origPath = origPath;
+    this.wasmModule = wasmModule;
+    this.allocateFn = allocateFn;
+  }
+
+  @Override
+  public boolean isImmutable() {
+    return true;
+  }
+
+  @Override
+  public void repr(Printer printer) {
+    printer.append("<wasm_module path=");
+    printer.repr(origPath);
+    printer.append(">");
+  }
+
+  public StarlarkPath getPath() {
+    return path;
+  }
+
+  @StarlarkMethod(
+      name = "path",
+      structField = true,
+      doc = "The path this WebAssembly module was loaded from.")
+  public Object getOrigPath() {
+    return origPath;
+  }
+
+  public StarlarkWasmExecutionResult execute(
+      String functionName,
+      byte[] input,
+      Duration timeout,
+      long memLimitBytes)
+      throws EvalException, InterruptedException {
+    Profiler prof = Profiler.instance();
+    try (SilentCloseable c = prof.profile(WASM_EXEC, () -> "execute " + functionName)) {
+      var memLimits = getMemLimits(memLimitBytes);
+      // Perform initialization and execution in a separate thread so it can be interrupted
+      // in case of timeout.
+      var wasmThreadFactory =
+          Thread.ofPlatform().name(Thread.currentThread().getName() + "_wasm").factory();
+      StarlarkWasmExecutionResult result;
+      String errMessage;
+      try (var executor = Executors.newSingleThreadExecutor(wasmThreadFactory)) {
+        return executor.invokeAny(
+            ImmutableList.of(() -> run(functionName, input, memLimits)),
+            timeout.toMillis(),
+            TimeUnit.MILLISECONDS);
+      } catch (TimeoutException e) {
+        errMessage = String.format("Error executing %s: timed out", functionName);
+      } catch (ExecutionException e) {
+        errMessage = String.format("Error executing %s: %s", functionName, e.getCause().getMessage());
+      }
+      return StarlarkWasmExecutionResult.newErr(errMessage);
+    }
+  }
+
+  private StarlarkWasmExecutionResult run(String execFunc, byte[] input, MemoryLimits memLimits)
+      throws EvalException, InterruptedException {
+    Instance instance;
+    Profiler prof = Profiler.instance();
+    try {
+      instance = Instance.builder(wasmModule)
+          .withMemoryLimits(memLimits)
+          // Disable calling `_start()`, which is the entry point for WASI-style
+          // command modules.
+          .withStart(false)
+          // Chicory documentation recommends ByteArrayMemory for OpenJDK
+          // https://chicory.dev/docs/advanced/memory
+          .withMemoryFactory(limits -> { return new ByteArrayMemory(limits); })
+          .build();
+      // If `_initialize()` is present then call it to perform early setup.
```

Comment:
seems to me that we should just let the usual "start function" take over this role?

Reply:
(see earlier comment)